### PR TITLE
Updates codebase to use Flax Linen & adds Mixer.

### PR DESCRIPTION
**Pull request by andsteing**
Main changes:

1. The entire codebase is refactored to use Flax Linen API & `ml_collections.ConfigDict` for experiment configuration.
2. MLP-Mixer is added (commits adding MLP-Mixer by @akolesnikoff), Colab and README updated accordingly.
3. Generation of new results with updated codebase & reference to runs in README. Also new default dropout setting for some models for increased performance.

### CODE REVIEW

File: vit_jax/train.py [hunk 1]
```diff
@@ -13,195 +13,215 @@
...
-      input_pipeline.DATASET_PRESETS[args.dataset]['total_steps'])
-
   # Prepare the learning-rate and pre-fetch it to device to avoid delays.
-  lr_fn = hyper.create_learning_rate_schedule(total_steps, args.base_lr,
-                                              args.decay_type,
-                                              args.warmup_steps)
-  lr_iter = hyper.lr_prefetch_iter(lr_fn, 0, total_steps)
-  update_rngs = jax.random.split(
-      jax.random.PRNGKey(0), jax.local_device_count())
+  update_rng_repl = flax.jax_utils.replicate(jax.random.PRNGKey(0))
```
Comment (lucasb-eyer) [inline vit_jax/train.py:152]: This changes the semantics from different seed per device to the same seed on all devices, which I would expect may change results!

Comment (andsteing) [reply vit_jax/train.py:152]: Note code inside `update_fn()`:

```
    _, new_rng = jax.random.split(rng)
    # Bind the rng key to the device id (which is unique across hosts)
    # Note: This is only used for multi-host training (i.e. multiple computers
    # each with multiple accelerators).
    dropout_rng = jax.random.fold_in(rng, jax.lax.axis_index('batch'))
```

So while the training loop pattern is different, the random values for dropout should still be unique to every host as before, PTAL.

Comment (akolesnikoff) [reply vit_jax/train.py:152]: looks good to me: we in fact switched to the exact same design internally.

**Review by akolesnikoff — APPROVED** [review]

File: vit_jax/train.py (hunk 1 above)
Comment (lucasb-eyer) [reply vit_jax/train.py:152]: Sounds good, I did scroll down to check that, but seems like I completely missed it!
